### PR TITLE
Support vim for FUSE on macOS

### PR DESCRIPTION
**Pull request by JiamingMai**
This PR does the following things to support `vim` for **FUSE** on **macOS**:
1. add renameCallback for macOS
2. support random write based on @elega 's commit

### CODE REVIEW

File: dora/integration/jnifuse/fs/src/main/java/alluxio/jnifuse/AbstractFuseFileSystem.java [hunk 1]
```diff
@@ -280,6 +280,15 @@ public int mkdirCallback(String path, long mode) {
...
   }
 
+  public int renameCallback(String oldPath, String newPath) {
+    try {
+      return rename(oldPath, newPath, 0);
+    } catch (Exception e) {
+      LOG.error("Failed to rename {}, newPath {}: ", oldPath, newPath, e);
+      return -ErrorCodes.EIO();
+    }
+  }
```
Comment (apc999) [inline dora/integration/jnifuse/fs/src/main/java/alluxio/jnifuse/AbstractFuseFileSystem.java:290]: Prefer `return renameCallback(oldPath, newPath, 0);` to remove code duplication

File: dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+
+/**
+ * foobar.
+ */
+public class FuseRandomAccessStream implements FuseFileStream {
```
Comment (apc999) [inline dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java:20]: nit: `RandomAccessFuseFileStream`

File: dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      }
+      out.close();
+      mRandomAccessFile = new RandomAccessFile(mTmpFile, "rw");
+      mURI = uri;
+      is.close();
+    } catch (Exception e) {
+      e.printStackTrace();
+      throw new RuntimeException(e);
+    }
+  }
```
Comment (apc999) [inline dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java:56]: potentially heavy IO and potential exception can be thrown in constructor here, which is anti-pattern.
let's use factory method instead for the IO part. 
always prefer making constructor light and succeed for sure.

File: dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+  @Override
+  public void flush() {
+    try {
+      // TODO(JiamingMai): rename it to a tmp file before deletion
+      mFileSystem.delete(mURI);
+      FileOutStream fos = mFileSystem.createFile(mURI);
```
Comment (apc999) [inline dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java:98]: always prefer `try-with-resource` to ensure corresponding resources are closed if an exception is thrown before close

File: dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+      long pos = mRandomAccessFile.getFilePointer();
+      mRandomAccessFile.seek(0);
+      FileChannel channel = mRandomAccessFile.getChannel();
+      ByteBuffer buf = ByteBuffer.allocate(COPY_TO_LOCAL_BUFFER_SIZE_DEFAULT);
```
Comment (apc999) [inline dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java:103]: allocate 64MB on heap can be very slow under heavy load.

**Review by apc999 — CHANGES_REQUESTED** [review]
a bunch of comments left after a quick review. thanks for the contribution!

File: dora/integration/jnifuse/fs/src/main/java/alluxio/jnifuse/AbstractFuseFileSystem.java (hunk 1 above)
Comment (JiamingMai) [reply dora/integration/jnifuse/fs/src/main/java/alluxio/jnifuse/AbstractFuseFileSystem.java]: Good catch. Updated.

File: dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java (hunk 2 above)
Comment (JiamingMai) [reply dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java]: Renamed.

File: dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java (hunk 5 above)
Comment (JiamingMai) [reply dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java]: Changed to 1MB.

File: dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java (hunk 3 above)
Comment (JiamingMai) [reply dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java]: Extracted this part of the logic as a method (`initTmpFile`) for lazy initialization.

File: dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java (hunk 4 above)
Comment (JiamingMai) [reply dora/integration/fuse/src/main/java/alluxio/fuse/file/FuseRandomAccessStream.java]: Updated. I split the method into two parts, and used `try-with-resource` to cover the `FileOutStream`.

**Pull request closed without merging**